src/app/clusters/operational-credentials-server/operational-credentials-server.cpp

```diff
@@ -238,6 +238,30 @@ void emberAfPluginOperationalCredentialsServerInitCallback(void)
     writeFabricsIntoFabricsListAttribute();
 }
 
+namespace {
+class OpCredsClusterExchangeDelegate : public Messaging::ExchangeDelegate
+{
+public:
+    CHIP_ERROR OnMessageReceived(Messaging::ExchangeContext * ec, const PacketHeader & packetHeader,
+                                 const PayloadHeader & payloadHeader, System::PacketBufferHandle && payload) override
+    {
+        return CHIP_NO_ERROR;
+    }
+    void OnResponseTimeout(Messaging::ExchangeContext * ec) override {}
+    void OnExchangeClosing(Messaging::ExchangeContext * ec) override
+    {
+        ec->GetExchangeMgr()->GetReliableMessageMgr()->ClearRetransTable(static_cast<Messaging::ReliableMessageContext *>(ec));
+        ec->GetExchangeMgr()->GetSessionMgr()->ExpireAllPairingsForFabric(mFabricIndex);
+        mFabricIndex = 0;
+    }
+
+    FabricIndex mFabricIndex = 0;
```

Comment:
Can't we retrieve the fabric index from the exchange context?  Why do we need to pre-store fabric index in what will be used for a global object?  This statefulness seems fragile given that this object is shared.

Reply:
The `ClearRetransTable` was also added due to a crash. `ReliableMessageMgr` was trying to access the exchange which was already closed. I added a fix in `ReliableMessageMgr` in the current PR since this change was made. Let me see if that change also addressed the crash that resulted into this code change. If so, I'll remove it from here.

Reply:
> This is the fabric index of the fabric being removed, which is not necessarily the same as the fabric index of the fabric associated with the exchange (accessing fabric), right?

I guess I was assuming that you can't remove a fabric that is not yours.  But maybe this isn't the right place to make that assumption.

In that case, my more direct critique is that we are staging state in the mFabricIndex field of this global object.  It seems like it should be possible to pass that state through the callback path instead of staging it ahead of time.  This is the type of thing that locks us into singleton patterns forever.

Reply:
I updated the code flow for the scenario where its not the local fabric. For local fabric, it'll get the index from the exchange context, as you had suggested.

Reply:
Looks good. Thanks.